servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ProxyConnectChannelSingle.java

```diff
@@ -0,0 +1,191 @@
+
+        private final ConnectionObserver observer;
+        private final HttpHeadersFactory headersFactory;
+        private final String connectAddress;
+        @Nullable
+        private Subscriber<? super Channel> subscriber;
+        @Nullable
+        private ProxyConnectObserver connectObserver;
+        @Nullable
+        private HttpResponseMetaData response;
+
+        private ProxyConnectHandler(final ConnectionObserver observer,
+                                    final HttpHeadersFactory headersFactory,
+                                    final String connectAddress,
+                                    final Subscriber<? super Channel> subscriber) {
+            this.observer = observer;
+            this.headersFactory = headersFactory;
+            this.connectAddress = connectAddress;
+            this.subscriber = subscriber;
+        }
+
+        @Override
+        public void handlerAdded(final ChannelHandlerContext ctx) {
+            if (ctx.channel().isActive()) {
+                sendConnectRequest(ctx);
+            }
+        }
+
+        @Override
+        public void channelActive(final ChannelHandlerContext ctx) {
+            sendConnectRequest(ctx);
+            ctx.fireChannelActive();
+        }
+
+        private void sendConnectRequest(final ChannelHandlerContext ctx) {
+            final HttpRequestMetaData request = newRequestMetaData(HTTP_1_1, CONNECT, connectAddress,
+                    headersFactory.newHeaders()).addHeader(HOST, connectAddress);
+            connectObserver = observer.onProxyConnect(request);
+            ctx.writeAndFlush(request).addListener(f -> {
+                if (f.isSuccess()) {
+                    ctx.read();
+                } else {
+                    failSubscriber(ctx, new RetryableProxyConnectException(
+                            "Failed to write CONNECT request", f.cause()));
+                }
+            });
+        }
+
+        @Override
+        public void channelRead(final ChannelHandlerContext ctx, final Object msg) {
+            if (msg instanceof HttpResponseMetaData) {
+                if (response != null) {
+                    failSubscriber(ctx, new RetryableProxyConnectException(
+                            "Received two responses for a single CONNECT request"));
+                    return;
+                }
+                response = (HttpResponseMetaData) msg;
+                if (response.status().statusClass() != SUCCESSFUL_2XX) {
+                    failSubscriber(ctx, new ProxyResponseException("Non-successful response '" + response.status() +
+                            "' from proxy on CONNECT " + connectAddress, response.status()));
+                }
+            }
+        }
+
+        @Override
+        public void channelReadComplete(final ChannelHandlerContext ctx) throws Exception {
+            if (subscriber != null) {
+                ctx.read(); // Keep requesting until finished
```

Comment:
Question for my understanding: why wouldn't we just run all the logic that is currently in `userEventTriggered` in here? (so, what's the benefit in waiting for the InboundDataEndEvent?)

Reply:
It's ideal to wait for the internal state machine of `HttpResponseDecoder` to complete before we notify the subscriber. Completion is signaled by `InboundDataEndEvent`. It helps us to make sure the response from the proxy is correct and does not include any payload body. If a malformed proxy responds with `content-length` header and the payload body, we will detect it as an error.
Will add error handling, a test, and a comment.

Reply:
Thanks!